File: src/workflow/adapters/base_adapter.py
"""
領域適配器基類 (Base Domain Adapter)

定義領域適配器的抽象接口，各領域 (Novel/Code/Knowledge) 
繼承此類實現具體的工作流邏輯。

設計原則:
- 抽象公共接口
- 強制實現必要方法
- 提供默認實現可覆蓋
"""

from abc import ABC, abstractmethod
from typing import Any, Dict, List, Optional, Type
from dataclasses import dataclass
from enum import Enum

from src.workflow.base_state import BaseState, BaseWorkflowConfig


# ============================================================
# 領域類型枚舉
# ============================================================

class DomainType(Enum):
    """領域類型"""
    NOVEL = "novel"
    CODE = "code"
    KNOWLEDGE = "knowledge"
    CUSTOM = "custom"


# ============================================================
# 評估結果基類
# ============================================================

@dataclass
class BaseEvaluationResult:
    """評估結果基類"""
    decision: str                   # APPROVED | REVISE | HUMAN_REVIEW | REWRITE
    decision_reason: str
    total_score: float
    dimension_scores: Dict[str, float]  # 各維度分數
    feedback: str                   # 可執行反饋
    revision_instructions: List[Dict[str, str]]  # 修改指令


# ============================================================
# 領域適配器基類
# ============================================================

class BaseDomainAdapter(ABC):
    """
    領域適配器抽象基類
    
    每個領域 (小說/代碼/知識) 需要實現:
    1. 狀態類型 (get_state_class)
    2. 驗證邏輯 (evaluate)
    3. 圖構建 (create_graph)
    4. 節點函數 (create_nodes)
    """
    
    def __init__(self, config: Optional[BaseWorkflowConfig] = None):
        self.config = config or {}
        self.domain = self.get_domain_type()
    
    # ========================================
    # 抽象方法 - 子類必須實現
    # ========================================
    
    @abstractmethod
    def get_domain_type(self) -> str:
        """返回領域類型標識"""
        pass
    
    @abstractmethod
    def get_state_class(self) -> Type[BaseState]:
        """返回該領域的狀態類"""
        pass
    
    @abstractmethod
    def create_initial_state(self, user_request: str, **kwargs) -> BaseState:
        """創建該領域的初始狀態 (允許額外元數據透傳)"""
        pass
    
    @abstractmethod
    def evaluate(self, state: BaseState) -> BaseEvaluationResult:
        """
        評估當前狀態質量
        
        各領域評估維度不同:
        - 小說: LOCK (Lead, Objective, Conflict, Knockout)
        - 代碼: Tests, Lint, Build, Coverage
        - 知識: Accuracy, Completeness, Relevance
        """
        pass
    
    @abstractmethod
    def create_graph(self):
        """
        創建 LangGraph 工作流圖
        
        Returns:
            StateGraph: LangGraph 狀態圖
        """
        pass
    
    # ========================================
    # 可覆蓋方法 - 提供默認實現
    # ========================================
    
    def get_nodes(self) -> Dict[str, callable]:
        """
        獲取工作流節點函數映射
        
        Returns:
            Dict[str, callable]: 節點名 -> 節點函數
        """
        return {}
    
    def get_routing_rules(self) -> Dict[str, callable]:
        """
        獲取路由規則映射
        
        Returns:
            Dict[str, callable]: 邊名 -> 路由函數
        """
        return {}
    
    def should_continue(self, state: BaseState) -> str:
        """
        默認的繼續/終止判斷
        
        可被子類覆蓋以實現領域特定邏輯
        """
        if state.get("revision_count", 0) >= state.get("max_revisions", 3):
            return "human_review"
        
        decision = state.get("decision", "")
        if decision == "APPROVED":
            return "finalize"
        elif decision == "HUMAN_REVIEW":
            return "human_review"
        elif decision in ("REVISE", "REWRITE"):
            return "revise"
        else:
            return "continue"
    
    def get_default_config(self) -> BaseWorkflowConfig:
        """獲取默認配置（基礎層 fallback；具體領域應覆蓋）"""
        return {
            # 僅適用於 base/custom 通用場景
            # novel 請以 src/workflow/state.py 的 domain 常量為準
            "pass_score": 80,
            "human_review_score": 70,
            "max_revisions": 3,
            "auto_approve_timeout": 300,
            "verbose": True,
            "save_intermediate": True,
            "domain": self.domain,
            "domain_config": {},
        }
    
    # ========================================
    # 工具方法
    # ========================================
    
    def merge_config(self, custom_config: Optional[Dict] = None) -> BaseWorkflowConfig:
        """合併配置"""
        default = self.get_default_config()
        if custom_config:
            default.update(custom_config)
        return default


# ============================================================
# 適配器註冊表
# ============================================================

class AdapterRegistry:
    """
    適配器註冊表

    用於動態註冊和查找領域適配器
    """

    _adapters: Dict[str, Type[BaseDomainAdapter]] = {}
    _adapter_capabilities: Dict[str, set[str]] = {}

    @staticmethod
    def _normalize_capabilities(capabilities: Optional[Any]) -> set[str]:
        if capabilities is None:
            return set()
        if isinstance(capabilities, str):
            values = [capabilities]
        else:
            values = list(capabilities)
        return {str(value).strip() for value in values if str(value).strip()}

    @classmethod
    def register(
        cls,
        domain: str,
        capabilities: Optional[Any] = None,
    ):
        """裝飾器: 註冊適配器"""

        def decorator(adapter_class: Type[BaseDomainAdapter]):
            cls.register_adapter(domain, adapter_class, capabilities=capabilities)
            return adapter_class

        return decorator

    @classmethod
    def register_adapter(
        cls,
        domain: str,
        adapter_class: Type[BaseDomainAdapter],
        capabilities: Optional[Any] = None,
    ) -> None:
        """註冊適配器類（支持 capability 元數據）"""
        cls._adapters[domain] = adapter_class
        cls._adapter_capabilities[domain] = cls._normalize_capabilities(capabilities)

    @classmethod
    def get(cls, domain: str) -> Optional[Type[BaseDomainAdapter]]:
        """獲取適配器類"""
        return cls._adapters.get(domain)

    @classmethod
    def get_capabilities(cls, domain: str) -> List[str]:
        """獲取領域能力標籤"""
        capabilities = cls._adapter_capabilities.get(domain, set())
        return sorted(capabilities)

    @classmethod
    def list_domains(cls) -> List[str]:
        """列出所有已註冊領域"""
        return list(cls._adapters.keys())

    @classmethod
    def list_domains_by_capability(cls, capability: str) -> List[str]:
        """按能力標籤列出領域"""
        needle = (capability or "").strip()
        if not needle:
            return []
        return [
            domain
            for domain in cls.list_domains()
            if needle in cls._adapter_capabilities.get(domain, set())
        ]

    @classmethod
    def create_adapter(cls, domain: str, config: Optional[Dict] = None) -> Optional[BaseDomainAdapter]:
        """創建適配器實例"""
        adapter_class = cls.get(domain)
        if adapter_class:
            return adapter_class(config)
        return None

File: src-ts/workflow/adapters/base-adapter.ts
/**
 * Base Domain Adapter - Abstract interface for domain adapters
 * and adapter registry.
 *
 * Merged from:
 *   src/workflow/base_adapter.py
 *   src/workflow/adapters/base_adapter.py
 *
 * Design principles:
 * - Abstract common interface
 * - Force implementation of required methods
 * - Provide default implementations that can be overridden
 */

import type { BaseState, BaseWorkflowConfig } from '../state.js';

// ============================================================
// Evaluation result interface
// ============================================================

export interface BaseEvaluationResult {
  /** APPROVED | REVISE | HUMAN_REVIEW | REWRITE */
  decision: string;
  decision_reason: string;
  total_score: number;
  /** Per-dimension scores */
  dimension_scores: Record<string, number>;
  /** Actionable feedback */
  feedback: string;
  /** Revision instructions */
  revision_instructions: Array<Record<string, string>>;
}

// ============================================================
// Node function type
// ============================================================

export type NodeFunction = (state: BaseState) => BaseState | Promise<BaseState>;
export type RoutingFunction = (state: BaseState) => string;

// ============================================================
// Stub interface for graph engine (replaces LangGraph StateGraph)
// ============================================================

export interface IWorkflowGraph {
  addNode(name: string, fn: NodeFunction): void;
  addEdge(fromNode: string, toNode: string): void;
  addConditionalEdge(fromNode: string, router: RoutingFunction, mapping: Record<string, string>): void;
  compile(): { invoke(state: BaseState): Promise<BaseState> };
}

// ============================================================
// Base domain adapter abstract class
// ============================================================

export abstract class BaseDomainAdapter {
  protected config: BaseWorkflowConfig;
  protected domain: string;

  constructor(config?: BaseWorkflowConfig | null) {
    this.config = config ?? {};
    this.domain = this.getDomainType();
  }

  // ========================================
  // Abstract methods - subclasses must implement
  // ========================================

  /** Return domain type identifier string. */
  abstract getDomainType(): string;

  /** Return the state class constructor for this domain. */
  abstract getStateClass(): new (...args: unknown[]) => BaseState;

  /** Create the initial state for this domain. */
  abstract createInitialState(userRequest: string, extra?: Record<string, unknown>): BaseState;

  /**
   * Evaluate current state quality.
   *
   * Dimensions differ per domain:
   * - Novel: LOCK (Lead, Objective, Conflict, Knockout)
   * - Code:  Tests, Lint, Build, Coverage
   * - Knowledge: Accuracy, Completeness, Relevance
   */
  abstract evaluate(state: BaseState): BaseEvaluationResult;

  /**
   * Create a workflow graph (LangGraph replacement).
   * Returns an IWorkflowGraph that can be compiled and invoked.
   */
  abstract createGraph(): IWorkflowGraph;

  // ========================================
  // Overridable methods - default implementations
  // ========================================

  /**
   * Get workflow node function mapping.
   * Returns: node name -> node function
   */
  getNodes(): Record<string, NodeFunction> {
    return {};
  }

  /**
   * Get routing rule mapping.
   * Returns: edge name -> routing function
   */
  getRoutingRules(): Record<string, RoutingFunction> {
    return {};
  }

  /**
   * Default continue/terminate logic.
   * Can be overridden by subclasses for domain-specific logic.
   */
  shouldContinue(state: BaseState): string {
    if ((state.revision_count ?? 0) >= (state.max_revisions ?? 3)) {
      return 'human_review';
    }

    const decision = state.decision ?? '';
    if (decision === 'APPROVED') {
      return 'finalize';
    } else if (decision === 'HUMAN_REVIEW') {
      return 'human_review';
    } else if (decision === 'REVISE' || decision === 'REWRITE') {
      return 'revise';
    } else {
      return 'continue';
    }
  }

  /**
   * Get default configuration (base-layer fallback;
   * specific domains should override).
   */
  getDefaultConfig(): BaseWorkflowConfig {
    return {
      pass_score: 80,
      human_review_score: 70,
      max_revisions: 3,
      auto_approve_timeout: 300,
      verbose: true,
      save_intermediate: true,
      domain: this.domain,
      domain_config: {},
    };
  }

  // ========================================
  // Utility methods
  // ========================================

  /** Merge custom config on top of defaults. */
  mergeConfig(customConfig?: Record<string, unknown> | null): BaseWorkflowConfig {
    const defaults = this.getDefaultConfig();
    if (customConfig) {
      Object.assign(defaults, customConfig);
    }
    return defaults;
  }
}

// ============================================================
// Adapter Registry
// ============================================================

type DomainAdapterConstructor = new (config?: BaseWorkflowConfig | null) => BaseDomainAdapter;

export class AdapterRegistry {
  private static _adapters: Record<string, DomainAdapterConstructor> = {};
  private static _adapterCapabilities: Record<string, Set<string>> = {};

  private static _normalizeCapabilities(capabilities: unknown): Set<string> {
    if (capabilities == null) return new Set();
    if (typeof capabilities === 'string') {
      const trimmed = capabilities.trim();
      return trimmed ? new Set([trimmed]) : new Set();
    }
    if (Array.isArray(capabilities)) {
      const result = new Set<string>();
      for (const value of capabilities as Iterable<unknown>) {
        const str = String(value).trim();
        if (str) result.add(str);
      }
      return result;
    }
    return new Set();
  }

  /**
   * Decorator-like: register an adapter class for a domain.
   * Usage: AdapterRegistry.registerAdapter('novel', NovelAdapter, ['writing'])
   */
  static registerAdapter(
    domain: string,
    adapterClass: DomainAdapterConstructor,
    capabilities?: unknown,
  ): void {
    AdapterRegistry._adapters[domain] = adapterClass;
    AdapterRegistry._adapterCapabilities[domain] = AdapterRegistry._normalizeCapabilities(capabilities);
  }

  /** Get the adapter class for a domain. */
  static get(domain: string): DomainAdapterConstructor | undefined {
    return AdapterRegistry._adapters[domain];
  }

  /** Get capability tags for a domain. */
  static getCapabilities(domain: string): string[] {
    const capabilities = AdapterRegistry._adapterCapabilities[domain] ?? new Set();
    return Array.from(capabilities).sort();
  }

  /** List all registered domains. */
  static listDomains(): string[] {
    return Object.keys(AdapterRegistry._adapters);
  }

  /** List domains that have a specific capability. */
  static listDomainsByCapability(capability: string): string[] {
    const needle = (capability ?? '').trim();
    if (!needle) return [];
    return AdapterRegistry.listDomains().filter(
      domain => AdapterRegistry._adapterCapabilities[domain]?.has(needle),
    );
  }

  /** Create an adapter instance for a domain. */
  static createAdapter(
    domain: string,
    config?: Record<string, unknown> | null,
  ): BaseDomainAdapter | null {
    const AdapterClass = AdapterRegistry.get(domain);
    if (AdapterClass) {
      return new AdapterClass(config as BaseWorkflowConfig);
    }
    return null;
  }
}
